File: lib/ping.js
/**
* Ping library, taken from node-ping and updated to return
* the durations of the ping time
*
* Thanks for the great module:
* (C) Daniel Zelisko
* http://github.com/danielzzz/node-ping
*
*/

var _child = require('child_process');
var _os = require('os');

var _platform = _os.platform();
var _path;
var _options;

if (_platform === 'linux') {
    _path = '/bin/ping';
    _options = ['-n', '-w 2', '-c 1'];
}
else if (_platform.match(/^win/)) {
    _path = 'C:/windows/system32/ping.exe';
    _options = ['-n', '1', '-w', '3000'];
}
else if (_platform === 'darwin') {
    _path = '/sbin/ping';
    _options = ['-n', '-t 2', '-c 1'];
}
else {
    console.error('Your platform is not supported.  We currently support Linux, Windows and OSX');
    process.exit(-1);
}

function pingHost(host, cb) {

    var output = "";
    var ping = null;

    var parameters = _options.slice(0);
    parameters.push(host);

    ping = _child.spawn(_path, parameters);

    ping.on('error', function(err) {
        var error = 'pingcheck: there was an error while executing the ping program. Please check the path or permissions.';
        error += '\npath: ' + _path;
        error += err.toString();
        return cb && cb(error);
    });

    ping.stdout.on('data', function (data) {
        output += String(data);
    });

    ping.stderr.on('data', function (data) {
        output += String(data);
    });

    ping.on('exit', function (code) {

        // if the there was an unknown host return that error
        if (/unknown host/i.test(output))
            return cb && cb('ERROR: The host "' + host + '" was not found');

        // if the there was an unknown host return that error (windows error message)
        if (/could not find host/i.test(output))
            return cb && cb('ERROR: The host "' + host + '" was not found');

        // if there was no output, there was an error, mark it as failed
        if (!output)
            return cb && cb(null);

        // if the process did not exit correctly, the host is down
        if (code !== 0)
            return cb && cb(null);

        // parse the data to get the ping time
        var pingTime;
        output.split('\n').forEach(function(line)
        {
            var time = line.match(/time=([0-9]*\.?[0-9]+)/);
            if (time) {
                pingTime = parseFloat(time[1]);
                return false;
            }
        });
        return cb && cb(null, pingTime);
    });
}

exports.ping = pingHost;

File: modules/ping.lua
local _child = require('childprocess')
local _os = require('los')
local uv = require('uv')

local _platform = _os.type()
local _path
local _options

if (_platform == 'linux') then

    _path = '/bin/ping'
    _options = {'-n', '-w 2', '-c 1'}

elseif (_platform == 'win32' ) then
    _path = 'C:/windows/system32/ping.exe'
    _options = {'-n', '1', '-w', '3000'}

elseif (_platform == 'darwin') then
    _path = '/sbin/ping'
    _options = {'-n', '-t 2', '-c 1'}

else 
    io.stderr:write('Your platform is not supported.  We currently support Linux, Windows and OSX')
    os.exit(-1)
end

function pingHost(host, cb)

    local output = ""
    local ping = nil

    table.insert(_options, host)
    ping = _child.spawn(_path, _options)
    table.remove(_options)

    ping:on('error', function(err)

        local error = 'pingcheck: There was an error while executing the ping program. Please check the path or permissions.'
        error = error .. '\npath: ' .. _path
        error = error .. '\n' .. tostring(err)
        
        return cb and cb(error)

    end)

    ping.stdout:on('data', function (data)
        output = output .. tostring(data)
    end)

    ping.stderr:on('data', function (data)
        output = output .. tostring(data) 
    end)

    ping:on('exit', function (code)        

        -- If the there was an unknown host return that error
        if ( string.find(output, "unknown host") ) then
            return cb and cb('ERROR: The host "' .. host .. '" was not found')
        end

        -- If the there was an unknown host return that error (windows error message) 
        if ( string.find(output, "could not find host") ) then
            return cb and cb('ERROR: The host "' .. host .. '" was not found')
        end

        -- If there was no output, there was an error, mark it as failed
        if (output == "") then
            return cb and cb('ERROR: Unable to obtain any output')
        end

        -- If the process did not exit correctly, the host is down
        if (code ~= 0) then
            return cb and cb('ERROR: Exit code is ' .. code)
        end

        -- Parse the data to get the ping time
        local pingTime, index, prevIndex, i, j, time = 0, 0, 0, 0, 0, 0

        while true do

            prevIndex = index
            index = string.find(output, "\n", index+1)

            if index == nil then break end

            local line = string.sub(output, prevIndex, index-1)
            i, j, time  = string.find(line, "time=([0-9]*%.?[0-9]+)")

            if(time) then 
                pingTime = tonumber(time)
                break
            end

        end        

        return cb and cb(nil, pingTime)        

    end)

end

exports.ping = pingHost